tests/components/mqtt/test_init.py

```diff
@@ -64,619 +58,642 @@ def mock_mqtt():
         yield mock_mqtt
 
 
-async def async_mock_mqtt_client(hass, config=None):
-    """Mock the MQTT paho client."""
-    if config is None:
-        config = {mqtt.CONF_BROKER: "mock-broker"}
+@pytest.fixture
+def calls():
+    """Fixture to record calls."""
+    return []
 
-    with patch("paho.mqtt.client.Client") as mock_client:
-        mock_client().connect.return_value = 0
-        mock_client().subscribe.return_value = (0, 0)
-        mock_client().unsubscribe.return_value = (0, 0)
-        mock_client().publish.return_value = (0, 0)
-        result = await async_setup_component(hass, mqtt.DOMAIN, {mqtt.DOMAIN: config})
-        assert result
-        await hass.async_block_till_done()
-        return mock_client()
 
+@pytest.fixture
+def record_calls(calls):
+    """Fixture to record calls."""
 
-mock_mqtt_client = threadsafe_coroutine_factory(async_mock_mqtt_client)
+    @callback
+    def record_calls(*args):
+        """Record calls."""
+        calls.append(args)
 
+    return record_calls
 
-# pylint: disable=invalid-name
-class TestMQTTComponent(unittest.TestCase):
-    """Test the MQTT component."""
 
-    def setUp(self):  # pylint: disable=invalid-name
-        """Set up things to be run when tests are started."""
-        self.hass = get_test_home_assistant()
-        mock_mqtt_component(self.hass)
-        self.calls = []
-        self.addCleanup(self.tear_down_cleanup)
+async def test_mqtt_connects_on_home_assistant_mqtt_setup(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test if client is connected after mqtt init on bootstrap."""
+    assert mqtt_client_mock.connect.call_count == 1
 
-    def tear_down_cleanup(self):
-        """Stop everything that was started."""
-        self.hass.stop()
 
-    @callback
-    def record_calls(self, *args):
-        """Record calls."""
-        self.calls.append(args)
+async def test_mqtt_disconnects_on_home_assistant_stop(hass, mqtt_mock):
+    """Test if client stops on HA stop."""
+    hass.bus.fire(EVENT_HOMEASSISTANT_STOP)
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    assert mqtt_mock.async_disconnect.called
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5737505)